crates/sui-core/src/authority_server.rs

```diff
@@ -426,56 +489,65 @@ impl ValidatorService {
                 .start_timer()
         };
 
-        // 1) Check if cert already executed
-        let tx_digest = *certificate.digest();
-        if let Some(signed_effects) = self
-            .state
-            .get_signed_effects_and_maybe_resign(&tx_digest, epoch_store)?
-        {
-            let events = if request.include_events {
-                if let Some(digest) = signed_effects.events_digest() {
-                    Some(self.state.get_transaction_events(digest)?)
+        // 1) Check if the certificate is already executed.
+        //    This is only needed when we have only one certificate (not a soft bundle).
+        if !is_soft_bundle {
```

Comment:
Given that you have already checked `is_tx_already_executed` previously, this check is not necessary.
You could simply use the first certificate for this, and if it's already there you return it.

Reply:
There may be a race if some other thread submitted the same tx right after the `is_tx_already_executed` check, resulting in the first certificate already been executed.

In this case, we will return with only one response (for the first certificate), while in a Soft Bundle context this is incorrect - because from protocol definition, the number of responses must match the number of requests.  The expect behavior, however, is to either assume nothing have happened and submit all these certificates again (that would probably lead to wrong ordering on chain, but acceptable), or just throw an error upfront.

Reply:
Would it be sufficient to do:
```
if certificates.len() == 1
```
instead of
```
if !is_soft_bundle
```
?